src/ccl_power.c

```diff
@@ -0,0 +1,1690 @@
+    if(cosmo->config.baryons_power_spectrum_method==ccl_bcm){
+      int pwstatus=0;
+      double fbcm=ccl_bcm_model_fkz(cosmo,k,a,&pwstatus);
+      pk=pk*fbcm;
+      if(pwstatus){
+        *status = CCL_ERROR_SPLINE_EV;
+        sprintf(cosmo->status_message ,"ccl_power.c: ccl_nonlin_matter_power(): Error in BCM correction\n");
+        return NAN;
+      }
+    }
+    return pk;
+
+  case ccl_emu:
+    if ((cosmo->config.transfer_function_method == ccl_emulator) && (a<A_MIN_EMU)){
+      *status = CCL_ERROR_EMULATOR_BOUND;
+      sprintf(cosmo->status_message, "ccl_power.c: the cosmic emulator cannot be used above z=2\
+\n");
+      return NAN;
+    }
+    
+    // Compute power spectrum if needed; return if computation failed
+    if (!cosmo->computed_power){
+      ccl_cosmology_compute_power(cosmo,status);
+    }
+    if (cosmo->data.p_nl == NULL) return NAN;
+    
+    if(k<=cosmo->data.k_min_nl) {
+      log_p_1=ccl_power_extrapol_lowk(cosmo,k,a,cosmo->data.p_nl,cosmo->data.k_min_nl,status);
+      return exp(log_p_1);
+    }
+    
+    if(k<cosmo->data.k_max_nl){
+      int gslstatus =  gsl_spline2d_eval_e(cosmo->data.p_nl, log(k),a,NULL ,NULL ,&log_p_1);
+      if(gslstatus != GSL_SUCCESS) {
+        ccl_raise_gsl_warning(gslstatus, "ccl_power.c: ccl_nonlin_matter_power():");
+        *status = CCL_ERROR_SPLINE_EV;
+        sprintf(cosmo->status_message ,"ccl_power.c: ccl_nonlin_matter_power(): Spline evaluation error\n");
+        return NAN;
+      }
+      else {
+	      pk = exp(log_p_1);
+	    }
+    }
+    else { // Extrapolate NL regime using log derivative
+      log_p_1 = ccl_power_extrapol_highk(cosmo,k,a,cosmo->data.p_nl,cosmo->data.k_max_nl,status);
+      pk = exp(log_p_1);
+    }
+    // Add baryonic correction
+    if(cosmo->config.baryons_power_spectrum_method==ccl_bcm){
+      int pwstatus=0;
+      double fbcm=ccl_bcm_model_fkz(cosmo,k,a,&pwstatus);
+      pk = pk*fbcm;
+      if(pwstatus){
+	    *status = CCL_ERROR_SPLINE_EV;
+	    sprintf(cosmo->status_message ,"ccl_power.c: ccl_nonlin_matter_power(): Error in BCM correction\n");
+	    return NAN;
+      }
+    }
+    return pk;
+    
+  default:
+    printf("WARNING:  config.matter_power_spectrum_method = %d not yet supported\n continuing with linear power spectrum\n",cosmo->config.matter_power_spectrum_method);
+    cosmo->config.matter_power_spectrum_method=ccl_linear;
+    return ccl_linear_matter_power(cosmo,k,a,status);
+  } // end switch
+
+}
+
+//Params for sigma(R) integrand
+typedef struct {
+  ccl_cosmology *cosmo;
+  double R;
+  int* status;
+} SigmaR_pars;
+
+static double sigmaR_integrand(double lk,void *params)
+{
+  SigmaR_pars *par=(SigmaR_pars *)params;
+  
+  double k=pow(10.,lk);
+  double pk=ccl_linear_matter_power(par->cosmo,k, 1.,par->status);
+  double kR=k*par->R;
+  double w;
+  if(kR<0.1) {
+    w =1.-0.1*kR*kR+0.003571429*kR*kR*kR*kR
+      -6.61376E-5*kR*kR*kR*kR*kR*kR
+      +7.51563E-7*kR*kR*kR*kR*kR*kR*kR*kR;
```

Comment:
Not gcc 8.2 with -O3.  It didn't improve upon the original 24 flops.  cf. https://godbolt.org/g/Ymk96e

Clang 6.0.0 with -O3 got the original down to 17 flops, but still nearly double the more efficient version.

In general compilers are good at turning a series of operations into SIMD code and vectorizing things to improve memory management.  But not reorganizing math into an equivalent calculation with fewer flops.  It's on you to understand what the math means and write it in an efficient way for the compiler.

Reply:
godbolt shows gcc does substantially better if you pass -ffast-math; it's probably prevented from optimizing because of strict adherence to the ieee floating point standards (which icc probably just ignores).

Reply:
Interesting, thanks.